### PR TITLE
Track subsystem runtimes via usr

### PR DESCRIPTION
**Pull request by Absolucy**
## About The Pull Request

This adds a dummy "subsystem tracker" mob to every subsystem - which is set to `usr` during SS preinit, init and fire. (surely this can't go wrong at all)

If `usr` is a subsystem tracker during a runtime, it'll replace `usr: blahblahblah` in the runtime text with `subsystem: SSwhatever (/datum/controller/subsystem/whatever)`

Also, this makes a `/mob/abstract` type, and repaths `/mob/oranges_ear` to `/mob/abstract/oranges_ear`, so that it can share the "technically this is mob but it's not *really* a mob" behavior with the dummy tracker mob.

blame @MrStonedOne

## Why It's Good For The Game

makes runtimes slightly easier to read

2024-10-24 (1729820355) ~ dreamseeker

## Changelog

... this isn't user facing at all.

### CODE REVIEW

**Comment by tgstation-server** [comment]
#### Test Merge Deployed

<details><summary>Details</summary>

##### Server Instance
Terry

##### Merged By
Auto\mrstonedone

##### Revision
Origin: 26361ae5656fd4241634ffead9f31b7d651677b9
Pull Request: 59455f63a9b0b8c2f8a95511c5acc7e0ebc5729d
Server: 747a6fa0635313afa6f6b0f24a01e202f7288936

##### Comment
Lets test a thing
[GitHub Deployments](https://github.com/tgstation/tgstation/deployments/activity_log?environment=TGS%3A+Terry)
</details>

**Comment by Absolucy** [comment]
why would you-

**Comment by tgstation-server** [comment]
#### Test Merge Updated

<details><summary>Details</summary>

##### Server Instance
Terry

##### Merged By
Auto\mrstonedone

##### Revision
Origin: cded23b51a2264d636a336bd645f4714bdb1e82e
Pull Request: 59455f63a9b0b8c2f8a95511c5acc7e0ebc5729d
Server: e03295d3beea8cfb4da573956b9377761107206c

##### Comment
Lets test a thing
</details>

**Comment by Thunder12345** [comment]
> why would you-

Guess what, Terry wouldn't restart by himself at round end after having a draft PR implanted in his brain

**Comment by tgstation-server** [comment]
#### Test Merge Removed

**Comment by MrStonedOne** [comment]
> why would you-

I had to test the new tgsv6 server's fancy features some how.

**Comment by Absolucy** [comment]
that's fair

**Comment by Absolucy** [comment]
lmao this might actually work

File: code/controllers/tracker.dm [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/// An abstract mob used for tracking which subsystem caused a runtime.
+/mob/abstract/subsystem_tracker
+
+// This needs to do its thing as soon as possible. We do not care about this thing being tracked as an atom,
+// because they will exist *before* SSatoms.
+/mob/abstract/subsystem_tracker/New(datum/controller/subsystem/subsystem)
```
Comment (ZephyrTFA) [inline code/controllers/tracker.dm:6]:
```suggestion
/mob/abstract/subsystem_tracker/New(loc, datum/controller/subsystem/subsystem)
```

File: code/controllers/tracker.dm [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+	if(!istype(subsystem))
```
Comment (ZephyrTFA) [inline code/controllers/tracker.dm:7]:
```suggestion
	if(loc != null)
		CRASH("Tried to create a subsystem tracker outside of null space!")
	if(!istype(subsystem))
```

File: code/__DEFINES/MC.dm [hunk 3]
```diff
@@ -87,65 +90,89 @@
...
+	if(!isnull(_tracker)) { \
+		usr = _old_usr; \
+	}
+
+
+#ifndef UNIT_TESTS
+/// Sets up the abstract usr tracker for the subsystem, and uses it for PreInit()
+#define SETUP_SS_TRACKER_AND_PREINIT \
+	if(!(flags & SS_NO_TRACKER)) { \
+		src.tracker = new(src); \
```
Comment (ZephyrTFA) [inline code/__DEFINES/MC.dm:110]:
```suggestion
		src.tracker = new(null, src); \
```

**Review by ZephyrTFA — CHANGES_REQUESTED** [review]

**Comment by Absolucy** [comment]
i don't care enough tbh. anyone else is fully free to pick this up if they care to

**Pull request closed without merging**